### PR TITLE
[IMP] mail: improve email previews

### PR DESCRIPTION
**Pull request by flch-odoo**
Prepare an informative and nice-looking preview from the main content of the
email body, avoiding buttons/images alt etc.
Links, images, tables are removed.

Whitespace is added after the preview to avoid including the full message in
the preview (with markup).

Tags and attributes are also added to increase compliance with HTML5 standard,
including accessibility.

Task-2413355

### CODE REVIEW

**Comment by flch-odoo** [comment]
Not sure if the test is at the right place or if they should be more closely linked to mail.message or notifications specifically.
Also, it would probably be nice to have a test for the notification template rendering of the preview with action buttons etc.

File: addons/mail/data/mail_templates_email_layouts.xml [hunk 1]
```diff
@@ -222,6 +236,27 @@
...
+        <template id="preview_with_padding">
+<div style="display: none; max-height: 0px; overflow: hidden;">
+    <t t-if="tracking_values">
+        <t t-esc="tracking_values[0][0]"/>: <t t-esc="tracking_values[0][1]"/> -&gt; <t t-esc="tracking_values[0][2]"/>
+        <t t-if="len(tracking_values) > 1"> |...</t>
+        <t t-if="message.hidden_preview"> | </t>
+    </t>
+    <t t-out="message.hidden_preview"/>
+<!-- Whitespace to push back email content so that it doesn't appear in preview. Specific characters to use are susceptible to change over time -->
+&#847; &#847; &#847; &#847; &#847; &#847; &#847; &#847; &#847; &#847; &#847; &#847; &#847; &#847; &#847; &#847; &#847; &#847; &#847; &#847;
```
Comment (std-odoo) [inline addons/mail/data/mail_templates_email_layouts.xml:251]: can you do something like this ?
```xml
<t t-out"'&#847;' * 100"/>
```

Comment (std-odoo) [reply addons/mail/data/mail_templates_email_layouts.xml:251]: I think it's better if you alternate white-space with zero width unicode `&#8203;`
I read on a blog that some mail client trim the text to build the preview (so this character break the trim)

File: addons/mail/data/mail_templates_email_layouts.xml [hunk 2]
```diff
@@ -2,6 +2,10 @@
 <odoo>
     <data>
         <template id="message_notification_email">
+<html t-att-lang="lang"><header>
```
Comment (std-odoo) [inline addons/mail/data/mail_templates_email_layouts.xml:5]: isn't this removed by the HTMl sanitizer ?

File: addons/mail/data/mail_templates_email_layouts.xml [hunk 3]
```diff
@@ -222,6 +236,27 @@
 </table>
 </td></tr>
 </table>
+</body></html>
+        </template>
+
+        <template id="preview_with_padding">
+<div style="display: none; max-height: 0px; overflow: hidden;">
+    <t t-if="tracking_values">
+        <t t-esc="tracking_values[0][0]"/>: <t t-esc="tracking_values[0][1]"/> -&gt; <t t-esc="tracking_values[0][2]"/>
```
Comment (std-odoo) [inline addons/mail/data/mail_templates_email_layouts.xml:245]: can you use t-out ?

File: odoo/tools/mail.py [hunk 4]
```diff
@@ -295,6 +298,29 @@ def html_keep_url(text):
     return final
 
 
+def html2text_only(html, sanitize_params=None, strip=True, keep_newlines=False, harmonize_whitespace=True):
```
Comment (std-odoo) [inline odoo/tools/mail.py:301]: I would remove all unused arguments

File: addons/mail/models/mail_message.py [hunk 5]
```diff
@@ -1032,6 +1040,17 @@ def _notify_message_notification_update(self):
     # TOOLS
     # ------------------------------------------------------
 
+    def _get_message_preview(self, max_char=190):
+        """Returns an unformatted version of the message body. Unless `max_char=0` is passed,
+        output will be capped at max_char characters with a ' [...]' suffix if applicable.
+        Default `max_char` is the longest known mail client preview length (Outlook 2013)."""
+        self.ensure_one()
+
+        plaintext_ct = html2text_only(self.body or '').strip()
```
Comment (std-odoo) [inline addons/mail/models/mail_message.py:1049]: IMO `html2text_only` should be able to receive False as a value, so you don't need this here

File: odoo/tools/mail.py [hunk 6]
```diff
@@ -260,6 +260,9 @@ def html_sanitize(src, silent=True, sanitize_tags=True, sanitize_attributes=Fals
 TEXT_URL_REGEX = r'https?://[a-zA-Z0-9@:%._\+~#=/-]+(?:\?\S+)?'
 # retrieve inner content of the link
 HTML_TAG_URL_REGEX = URL_REGEX + r'([^<>]*>([^<>]+)<\/)?'
+HTML2TEXT_NEWLINES = re.compile('<(div|p|br|tr)[^>]*>|\n')
+# From c24b @ https://stackoverflow.com/a/12982689/18408071 - CC BY-SA 4.0
+HTML2TEXT_STRIP_HTML = re.compile('<.*?>|&([a-z0-9]+|#[0-9]{1,6}|#x[0-9a-f]{1,6});')
```
Comment (std-odoo) [inline odoo/tools/mail.py:265]: I feel like we might have some corner cases where the HTML are not stripped, as you do it after the HTML sanitation, you can just do something like
```
HTML2TEXT_STRIP_HTML = re.compile('<.*?>')
```

(because all `<>` that does not belong to an HTML tag will be escaped)

**Review by std-odoo — COMMENTED** [review]
@flch-odoo Hi :) 

I have some questions / propositions :)

Cheers

**Review by flch-odoo — COMMENTED** [review]
Thank you for the comments, @std-odoo , most of which were translated into code changes. I'm ready for the next round, if applicable ;-)

File: odoo/tools/mail.py (hunk 6 above)
Comment (std-odoo) [reply odoo/tools/mail.py:265]: Ah ok I see, for which use case the second part is used ?
```
|&([a-z0-9]+|#[0-9]{1,6}|#x[0-9a-f]{1,6});
```

File: addons/mail/data/mail_templates_email_layouts.xml (hunk 2 above)
Comment (std-odoo) [reply addons/mail/data/mail_templates_email_layouts.xml:5]: I think it is, so is this content not sanitized ?

screenshot_spect

File: odoo/tools/mail.py (hunk 6 above)
Comment (flch-odoo) [reply odoo/tools/mail.py]: > it also covers characters that are not in <>, such as **`&lt;`**, which we don't want to see either. See test_html2text_only.

If you have a single `<` for example, sanitation will convert it to `&lt;`, which, while not a real problem in an email preview, it's not exactly 'text only" either :thinking:

I'll add this as a parameter, if you'll allow this one, because otherwise function name is misleading... and I don't think we should rename it to be only a `html2_email_preview`, right?

File: addons/mail/data/mail_templates_email_layouts.xml (hunk 2 above)
Comment (flch-odoo) [reply addons/mail/data/mail_templates_email_layouts.xml:5]: Well, I'm looking at the received emails (would you like a cc?). So, no, I suppose the rendered template is not sanitized in this way. Do you think it should, even if all the included contents should themselves be templates or have been sanitized?

Comment (std-odoo) [reply addons/mail/data/mail_templates_email_layouts.xml:5]: no I guess it's ok :) mail client are smart :)

File: odoo/tools/mail.py (hunk 6 above)
Comment (std-odoo) [reply odoo/tools/mail.py]: Mhh, isn't it better to just use the regex `<.*?>` and then html unescape the result ?

If you have some special character in the text they will be removed

File: odoo/tools/mail.py [hunk 7]
```diff
@@ -295,6 +297,20 @@ def html_keep_url(text):
     return final
 
 
+def get_html_content(html):
```
Comment (tde-banana-odoo) [inline odoo/tools/mail.py:300]: html2unformattedtext?

File: odoo/addons/base/tests/test_mail.py [hunk 8]
```diff
@@ -306,6 +306,17 @@ def test_plaintext2html(self):
             html = plaintext2html(content, container_tag)
             self.assertEqual(html, expected, 'plaintext2html is broken')
 
+    def test_get_html_content(self):
+        cases = [
```
Comment (tde-banana-odoo) [inline odoo/addons/base/tests/test_mail.py:310]: What happens with wrong html (missing tags, ...) ?

File: odoo/tools/mail.py [hunk 9]
```diff
@@ -295,6 +297,20 @@ def html_keep_url(text):
...
+    """Returns unformatted text after removing html tags and excessive whitespace from a
+    string/Markup. Passed strings will first be sanitized.
+    """
+    html = html or ''
```
Comment (tde-banana-odoo) [inline odoo/tools/mail.py:304]: Maybe we could make an early skip if content is void, like
```
if not html or is_html_empty(html):
    return ''
```

File: addons/test_mail/tests/test_message_post.py [hunk 10]
```diff
@@ -295,9 +295,10 @@ def test_notify(self):
         admin_mails = [mail for mail in self._mails if self.partner_admin.name in mail.get('email_to')[0]]
         self.assertEqual(len(admin_mails), 1, 'There should be exactly one email sent to admin')
         admin_mail_body = admin_mails[0].get('body')
-        admin_access_link = admin_mail_body[admin_mail_body.index('model='):admin_mail_body.index('/>') - 1] if 'model=' in admin_mail_body else None
 
-        self.assertIsNotNone(admin_access_link, 'The email sent to admin should contain an access link')
+        self.assertTrue('model=' in admin_mail_body, 'The email sent to admin should contain an access link')
+        admin_access_link = admin_mail_body[
```
Comment (tde-banana-odoo) [inline addons/test_mail/tests/test_message_post.py:300]: Why this change ?

File: addons/mail/data/mail_templates_email_layouts.xml [hunk 11]
```diff
@@ -222,6 +239,20 @@
 </table>
 </td></tr>
 </table>
+</body></html>
+        </template>
+
+        <template id="preview_with_padding">
```
Comment (tde-banana-odoo) [inline addons/mail/data/mail_templates_email_layouts.xml:245]: ``notification_preview`` ? Not sure the "with padding" is useful (if someday we change the padding we would have to change the name :D ).

File: addons/mail/data/mail_templates_email_layouts.xml [hunk 12]
```diff
@@ -222,6 +239,20 @@
...
+<div style="display: none; max-height: 0px; overflow: hidden;">
+    <t t-if="tracking_values">
+        <t t-out="tracking_values[0][0]"/>: <t t-out="tracking_values[0][1]"/> &#8594; <t t-out="tracking_values[0][2]"/>
+        <t t-if="len(tracking_values) > 1"> |...</t>
+        <t t-if="message.hidden_preview"> | </t>
+    </t>
+    <t t-if="subtype_internal">Internal communication: </t><t t-out="message.hidden_preview"/>
```
Comment (tde-banana-odoo) [inline addons/mail/data/mail_templates_email_layouts.xml:252]: Is the "internal communication" really adding something here ?

File: addons/mail/data/mail_templates_email_layouts.xml [hunk 13]
```diff
@@ -125,17 +136,23 @@
...
 </td></tr>
 </table>
+</body></html>
         </template>
 
         <!-- Used mainly in conjunction of portal models (SO, PO, Invoices) although not mandatory -->
         <template id="mail_notification_paynow" name="Mail: Pay Now mail notification template">
-<table border="0" cellpadding="0" cellspacing="0" style="background-color: #FFFFFF; font-family:Verdana, Arial,sans-serif; color: #454748; width: 100%; border-collapse:separate;"><tr><td>
-<table border="0" cellpadding="0" cellspacing="0" style="max-width:900px; width:100%; background-color: #FFFFFF; color: #454748; border-collapse:separate;">
+<html t-att-lang="lang"><header>
```
Comment (tde-banana-odoo) [inline addons/mail/data/mail_templates_email_layouts.xml:144]: Even if we don't indent, would multi line, like
```
<html t-att-lang="lang">
<header>
    <meta http-equiv="Content-Type" content="text/html charset=UTF-8" />
</header>
<body>
<t t-set="...>
...
</body>
</html>
```

**Review by tde-banana-odoo — COMMENTED** [review]
Kataklops !

File: addons/mail/models/mail_message.py [hunk 14]
```diff
@@ -1032,6 +1040,17 @@ def _notify_message_notification_update(self):
...
+    def _get_message_preview(self, max_char=190):
+        """Returns an unformatted version of the message body. Unless `max_char=0` is passed,
+        output will be capped at max_char characters with a ' [...]' suffix if applicable.
+        Default `max_char` is the longest known mail client preview length (Outlook 2013)."""
+        self.ensure_one()
+
+        plaintext_ct = get_html_content(self.body)
+        if not max_char or len(plaintext_ct) <= max_char:
+            return plaintext_ct
+        return plaintext_ct[:max_char] + ' [...]'
```
Comment (tde-banana-odoo) [inline addons/mail/models/mail_message.py:1052]: You may use ``textwrap.shorten``

File: odoo/tools/mail.py (hunk 7 above)
Comment (tde-banana-odoo) [reply odoo/tools/mail.py:300]: Mmmmh, but it is not returning valid html. get_html_content from "html" source, somehow I think I will get a valid html content, a parsed html content, ... here it is basically text with some entities. Not sure about the naming, but would update it ^^
